module/Organizations/src/Organizations/ImageFileCache/Manager.php

```diff
@@ -0,0 +1,156 @@
+<?php
+/**
+ * YAWIK
+ *
+ * @filesource
+ * @license    MIT
+ * @copyright  2013 - 2016 Cross Solution <http://cross-solution.de>
+ */
+namespace Organizations\ImageFileCache;
+
+use Organizations\Entity\OrganizationImage;
+use Organizations\Options\ImageFileCacheOptions as Options;
+use Zend\Stdlib\ErrorHandler;
+use InvalidArgumentException;
+use RuntimeException;
+
+/**
+ * Image file cache manager
+ *
+ * @author Miroslav Fedeleš <miroslav.fedeles@gmail.com>
+ * @since 0.28
+ */
+class Manager
+{
+    
+    /**
+     * @var Options
+     */
+    protected $options;
+    
+    /**
+     * @param Options $options
+     */
+    public function __construct(Options $options)
+    {
+        $this->options = $options;
+    }
+
+    /**
+     * Returns image URI
+     *
+     * @param OrganizationImage $image
+     * @return string
+     */
+    public function getUri(OrganizationImage $image)
+    {
+        if ($this->options->getEnabled()) {
+            return sprintf('%s/%s', $this->options->getUriPath(), $this->getImageSubPath($image));
+        }
+        
+        return $image->getUri();
+    }
+    
+    /**
+     * @return bool
+     */
+    public function isEnabled()
+    {
+        return $this->options->getEnabled();
+    }
+    
+    /**
+     * Store an image as a file in a file system
+     *
+     * @param OrganizationImage $image
+     */
+    public function store(OrganizationImage $image)
+    {
+        $resource = $image->getResource();
+        $path = $this->getImagePath($image);
+        $this->createDirectoryRecursively(dirname($path));
+        file_put_contents($path, $resource);
+    }
+    
+    /**
+     * Delete an image file from file system
+     *
+     * @param OrganizationImage $image
+     */
+    public function delete(OrganizationImage $image)
+    {
+        @unlink($this->getImagePath($image));
+    }
+
+    /**
+     * Match the passed $uri and return an image ID on success
+     *
+     * @param string $uri
+     * @return null|string Image ID
+     */
+    public function matchUri($uri)
+    {
+        $pattern = '#^' . preg_quote($this->options->getUriPath(), '#') . '/[0-9a-z]/[0-9a-z]/([0-9a-z]+)\.[a-zA-Z]{3,4}$#';
+        $matches = [];
+        preg_match($pattern, $uri, $matches);
+        
+        return isset($matches[1]) ? $matches[1] : null;
+    }
+    
+    /**
+     * @param OrganizationImage $image
+     * @return string
+     */
+    protected function getImagePath(OrganizationImage $image)
+    {
+        return sprintf('%s/%s', $this->options->getFilePath(), $this->getImageSubPath($image));
+    }
+    
+    /**
+     * @param OrganizationImage $image
+     * @return string
+     */
+    protected function getImageSubPath(OrganizationImage $image)
+    {
+        $id = $image->getId();
+        
+        if (!$id) {
+            throw new InvalidArgumentException('image must have ID');
+        }
+        
+        $extension = pathinfo($image->getName(), PATHINFO_EXTENSION);
```

Comment:
It could be that an image DOES NOT have a name (by submitting jobs via the Import API). In that case, the image name should be composed using the image id and the information from the mimetype (image/**gif**).

``` php
$name = $image->getName();
if (!$name) { 
    $mimetype = $image->getMimetype();
     if (!$mimetype) { // throw exception cannot compose name. }
    $name = $image->getId() . '.' . str_replace('image/', '', $image->getMimetype();
}
```
Later we should propably move this logic to the image entity post insert hook.